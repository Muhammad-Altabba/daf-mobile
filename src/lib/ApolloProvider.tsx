import React from 'react'
import { ActivityIndicator } from 'react-native'
import { ApolloProvider } from 'react-apollo'
import { ApolloProvider as ApolloHooksProvider } from '@apollo/react-hooks'
import { ApolloClient, Resolvers } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { from } from 'apollo-link'
import { SchemaLink } from 'apollo-link-schema'
import { makeExecutableSchema } from 'graphql-tools'
import { Container, Screen, Text } from '@kancha/kancha-ui'

import { core, dataStore, db, resolvers, typeDefs } from './setup'

export const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
})

const contextLink = new SchemaLink({
  schema,
  context: { core, dataStore },
})

const link = from([contextLink])

export const cache = new InMemoryCache({})

export const client = new ApolloClient({
  connectToDevTools: true,
  cache,
  link,
})

interface Props {}
interface State {
  isRunningMigrations: boolean
}

class CustomProvider extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      isRunningMigrations: true,
    }
  }

  async componentDidMount() {
    await db.initialize()
    await dataStore.initialize()
    await core.startServices()

    this.setState({ isRunningMigrations: false })
    await core.syncServices(await dataStore.latestMessageTimestamps())
  }

  render() {
    if (this.state.isRunningMigrations) {
      return (
        <Screen>
          <Container flex={1} alignItems={'center'} justifyContent={'center'}>
            <ActivityIndicator size={'large'} />
          </Container>
        </Screen>
      )
    } else {
      return (
        <ApolloProvider client={client}>
          <ApolloHooksProvider client={client}>
            {this.props.children}
          </ApolloHooksProvider>
        </ApolloProvider>
      )
    }
  }
}

export default CustomProvider
