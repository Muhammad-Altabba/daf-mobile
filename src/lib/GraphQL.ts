import { ApolloClient, Resolvers } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { from } from 'apollo-link'
import { HttpLink } from 'apollo-link-http'
import gql from 'graphql-tag'

// const httpLink = new HttpLink({uri: '/graphql'})

const link = from([
  // httpLink,
])

const resolvers: Resolvers = {
  Query: {},
  Mutation: {
    log: (_, { type, message, category }, context) => {
      const query = gql`
        query getLogs {
          logs @client {
            id
            timestamp
            type
            category
            message
          }
        }
      `

      const previous = context.cache.readQuery({ query })
      const timestamp = new Date().getTime() / 1000
      const logItem = {
        id: `LogMessage:${timestamp}`,
        timestamp,
        type,
        message,
        category,
        __typename: 'LogMessage',
      }
      const data = {
        logs: [logItem, ...previous.logs],
      }

      context.cache.writeQuery({ query, data })
      return logItem
    },
  },
}

export const cache = new InMemoryCache()
export const client = new ApolloClient({
  connectToDevTools: true,
  typeDefs: gql`
    type LogMessage {
      id: ID!
      timestamp: Int
      type: Int
      message: String
      category: String
    }
    extend type Query {
      logs: [LogMessage!]
    }
  `,
  cache,
  link,
  resolvers,
})

cache.writeData({
  data: {
    logs: [],
  },
})
