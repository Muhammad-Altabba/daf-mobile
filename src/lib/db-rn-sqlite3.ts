import { openDatabase } from 'react-native-sqlite-storage'
import { Types } from './serto-graph'

export class RnSqlite implements Types.DbDriver {
  private db: any

  initialize(): Promise<any> {
    let setDb = (db: any) => {
      this.db = db
    }
    setDb = setDb.bind(this)

    return new Promise((resolve, reject) => {
      const db = openDatabase(
        { name: 'test', location: 'default' },
        () => {
          setDb(db)
          resolve()
        },
        reject,
      )
    })
  }

  run(sql: string, params: any): Promise<any> {
    return new Promise((resolve, reject) => {
      this.db.transaction((tx: any) => {
        tx.executeSql(sql, params || [], (t, result) => {
          //todo reject
          resolve()
        })
      })
    })
  }

  rows(sql: string, params: any): Promise<any> {
    return new Promise((resolve, reject) => {
      this.db.transaction((tx: any) => {
        tx.executeSql(sql, params || [], (t, result) => {
          //todo reject
          resolve(result.rows.raw())
        })
      })
    })
  }
}
