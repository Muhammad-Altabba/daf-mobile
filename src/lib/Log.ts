import gql from "graphql-tag";
import { cache, client } from "./GraphQL";

export enum LogMessageType {
  Info,
  Warning,
  Error
}

export interface LogMessage {
  id: string;
  type: LogMessageType;
  category?: string;
  message: string;
  timestamp: number;
  __typename?: string;
}

export const getLogsQuery = gql`
  query getLogs {
    logs @client {
      id
      timestamp
      type
      category
      message
    }
  }
`;

const newLogItemMutation = gql`
  mutation newLogItem($message: String!, $type: Int!, $category: String) {
    log(message: $message, type: $type, category: $category) @client
  }
`;

const writeLogItemToCache = (
  message: string,
  type: LogMessageType,
  category?: string
) => {
  client.mutate({
    mutation: newLogItemMutation,
    variables: {
      message,
      type,
      category
    },
    refetchQueries: ["getLogs"]
  });
};

export default {
  info: (message: string, category?: string) => {
    console.log(category && `[${category}]`, message);
    writeLogItemToCache(message, LogMessageType.Info, category);
  },

  warning: (message: string, category?: string) => {
    console.log(category && `[${category}]`, "⚠️ Warning:", message);
    writeLogItemToCache(message, LogMessageType.Warning, category);
  },

  error: (message: string, category?: string) => {
    console.log(category && `[${category}]`, "🛑 Error:", message);
    writeLogItemToCache(message, LogMessageType.Error, category);
  }
};
