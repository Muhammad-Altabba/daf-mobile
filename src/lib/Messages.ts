import { SertoMessage } from './serto-credentials'
import { client } from './GraphQL'
import { Queries } from './serto-graph'

export const saveMessage = async (message: SertoMessage) => {
  console.log('HERE', message)
  try {
    await client.mutate({
      mutation: Queries.newMessage,
      variables: {
        json: JSON.stringify(message),
      },
      refetchQueries: ['FindMessages($iss: ID, $sub: ID)'],
    })
  } catch (e) {
    console.log(e)
  }
  return ''
}
