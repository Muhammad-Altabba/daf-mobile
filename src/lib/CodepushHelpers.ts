import codePush, { DownloadProgress } from 'react-native-code-push'
import Log from './Log'

export const handleCodePushStatusChange = (status: codePush.SyncStatus) => {
  switch (status) {
    case codePush.SyncStatus.CHECKING_FOR_UPDATE:
      Log.info('Checking for updates.', 'Codepush')
      break
    case codePush.SyncStatus.DOWNLOADING_PACKAGE:
      Log.info('Downloading package.', 'Codepush')
      break
    case codePush.SyncStatus.INSTALLING_UPDATE:
      Log.info('Installing update.', 'Codepush')
      break
    case codePush.SyncStatus.UP_TO_DATE:
      Log.info('Up-to-date.', 'Codepush')
      break
    case codePush.SyncStatus.UPDATE_INSTALLED:
      Log.info('Update installed.', 'Codepush')
      break
  }
}
