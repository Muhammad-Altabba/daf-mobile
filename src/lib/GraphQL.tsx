import React from 'react'
import { ApolloProvider } from 'react-apollo'
import { ApolloClient, Resolvers } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { from } from 'apollo-link'
import { SchemaLink } from 'apollo-link-schema'
import { Api, typeDefs, resolvers } from './serto-graph'
import { RnSqlite } from './db-rn-sqlite3'
import { makeExecutableSchema } from 'graphql-tools'
import Log from './Log'
import { Container, Screen, Text } from '@kancha/kancha-ui'
import { syncEdges } from './TGEClient'

const localTypeDefs = `

  type LogMessage {
    id: ID!
    timestamp: Int
    type: Int
    message: String
    category: String
  }
  extend type Query {
    logs: [LogMessage!]
  }
`

export const schema = makeExecutableSchema({
  typeDefs: typeDefs + localTypeDefs,
  resolvers,
})

// local
const driver = new RnSqlite()
const api = new Api(driver)

const did = ''
const viewer = {
  did,
  ownsDids: api.getOwnedDids(did),
  isAdmin: true,
}

const contextLink = new SchemaLink({
  schema,
  context: { api, viewer },
})

const link = from([contextLink])

export const cache = new InMemoryCache()
export const client = new ApolloClient({
  connectToDevTools: true,
  cache,
  link,
})

cache.writeData({
  data: {
    logs: [],
  },
})

interface Props {}
interface State {
  isRunningMigrations: boolean
}

class CustomProvider extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      isRunningMigrations: true,
    }
  }

  componentDidMount() {
    driver
      .initialize()
      .then(() => api.initialize())
      .then(() => {
        this.setState({ isRunningMigrations: false })
        syncEdges(client)
      })
      .catch(e => console.log(e))
  }

  render() {
    if (this.state.isRunningMigrations) {
      return (
        <Screen safeArea={true}>
          <Container flex={1} alignItems={'center'} justifyContent={'center'}>
            <Text>Updating data...</Text>
          </Container>
        </Screen>
      )
    } else {
      return (
        <ApolloProvider client={client}>{this.props.children}</ApolloProvider>
      )
    }
  }
}

export default CustomProvider
