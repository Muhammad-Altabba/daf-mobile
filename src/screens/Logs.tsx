/**
 * Serto Mobile App
 *
 */

import React, { useState } from 'react'
import { StyleSheet, Text, View, FlatList } from 'react-native'
import { Query } from 'react-apollo'
import { LogMessage, LogMessageType, getLogsQuery } from '../lib/Log'
import moment from 'moment'

interface Props {}

export default (props: Props) => {
  return (
    <Query query={getLogsQuery}>
      {({
        data,
        loading,
        refetch,
      }: {
        data: { logs: LogMessage[] }
        loading: boolean
        refetch: () => void
      }) => (
        <FlatList
          data={data.logs}
          renderItem={({ item }) => <ListItem logItem={item} />}
          keyExtractor={(item, index) => item.id}
          onRefresh={refetch}
          refreshing={loading}
        />
      )}
    </Query>
  )
}

const ListItem = ({ logItem }: { logItem: LogMessage }) => {
  return (
    <View
      style={[
        styles.item,
        logItem.type === LogMessageType.Info && styles.info,
        logItem.type === LogMessageType.Warning && styles.warning,
        logItem.type === LogMessageType.Error && styles.error,
      ]}
    >
      <View style={styles.header}>
        <Text style={styles.category}>{logItem.category}</Text>
        <Text style={styles.date}>
          {moment.unix(logItem.timestamp).calendar()}
        </Text>
      </View>
      <Text style={styles.message}>{logItem.message}</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  item: {
    marginLeft: 3,
    marginTop: 1,
    marginBottom: 1,
    borderLeftWidth: 3,
    padding: 10,
  },
  info: {
    borderColor: 'gray',
  },
  warning: {
    borderColor: 'orange',
  },
  error: {
    borderColor: 'red',
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  category: {
    fontSize: 12,
    color: 'gray',
    fontWeight: 'bold',
  },
  message: {
    fontSize: 15,
  },
  date: {
    fontSize: 12,
    marginBottom: 5,
    color: 'gray',
  },
})
