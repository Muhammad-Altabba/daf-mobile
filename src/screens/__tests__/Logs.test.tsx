import 'react-native'
import React from 'react'
import { render } from 'react-native-testing-library'
import { MockedProvider } from '@apollo/react-testing'
import Logs from '../settings/Logs'
import { LogMessage, LogMessageType, getLogsQuery } from '../../lib/Log'

jest.useFakeTimers()
jest.runAllTimers()

const mockLogItem: LogMessage = {
  message: 'Sample info',
  type: LogMessageType.Info,
  category: 'Example',
  timestamp: 1234567890,
  id: '1234567',
}

const mocks = [
  {
    request: {
      query: getLogsQuery,
      variables: {},
    },
    result: {
      data: {
        logs: [mockLogItem],
      },
    },
  },
]

test('renders correctly', () => {
  const tree = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Logs />
    </MockedProvider>,
  ).toJSON()

  expect(tree).toMatchSnapshot()
})
