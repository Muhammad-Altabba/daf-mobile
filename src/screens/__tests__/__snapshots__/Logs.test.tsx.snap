// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": 1,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <RCTScrollView
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={true}
        />
      }
      refreshing={true}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "backgroundColor": "#EFEEF3",
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
  </View>
</View>
`;
